vm/array.go

```diff
@@ -247,4 +267,144 @@ var builtinArrayMethods = []*BuiltInMethod{
 		},
 		Name: "select",
 	},
+	{
+		Fn: func(receiver Object) builtinMethodBody {
+			return func(vm *VM, args []Object, blockFrame *callFrame) Object {
+				i := args[0]
+				index, ok := i.(*IntegerObject)
+
+				if !ok {
+					return newError("Expect index argument to be Integer. got=%T", i)
+				}
+
+				arr := receiver.(*ArrayObject)
+
+				if len(arr.Elements) == 0 {
+					return NULL
+				}
+
+				if int(index.Value) >= len(arr.Elements) {
+					return newError("Index out of range")
+				}
+
+				return arr.Elements[index.Value]
+			}
+		},
+		Name: "at",
+	},
+	{
+		Fn: func(receiver Object) builtinMethodBody {
+			return func(vm *VM, args []Object, blockFrame *callFrame) Object {
+				arr := receiver.(*ArrayObject)
+				arr.Elements = []Object{}
+
+				return arr
+			}
+		},
+		Name: "clear",
+	},
+	{
+		Fn: func(receiver Object) builtinMethodBody {
+			return func(vm *VM, args []Object, blockFrame *callFrame) Object {
+				arr := receiver.(*ArrayObject)
+
+				for _, arg := range args {
+					addAr, ok := arg.(*ArrayObject)
+
+					if !ok {
+						return newError("Expect argument to be Array. got=%T", arg)
+					}
+
+					for _, el := range addAr.Elements {
+						arr.Elements = append(arr.Elements, el)
+					}
+				}
+
+				return arr
+			}
+		},
+		Name: "concat",
+	},
+	{
+		Fn: func(receiver Object) builtinMethodBody {
+			return func(vm *VM, args []Object, blockFrame *callFrame) Object {
+				arr := receiver.(*ArrayObject)
+				var count int
+
+				if blockFrame != nil {
+					for _, obj := range arr.Elements {
+						result := builtInMethodYield(vm, blockFrame, obj)
+						if result.Target.(*BooleanObject).Value {
+							count++
+						}
+					}
+
+					return initilaizeInteger(count)
+				}
+
+				if len(args) > 1 {
+					return newError("Expect one argument. got=%d", len(args))
+				}
+
+				if len(args) == 0 {
+					return initilaizeInteger(len(arr.Elements))
+				}
+
+				arg := args[0]
+				findInt, findIsInt := arg.(*IntegerObject)
+				findString, findIsString := arg.(*StringObject)
+				findBoolean, findIsBoolean := arg.(*BooleanObject)
+
+				for i := 0; i < len(arr.Elements); i++ {
+					elInt, isInt := arr.Elements[i].(*IntegerObject)
+					if isInt && findIsInt && elInt.Value == findInt.Value {
```

Comment:
I think you should write some helper functions for these type checking and comparison operations. You can use them in other array's methods or hash's methods.